prog/postgres/postgres_server_nexus.rb

```diff
@@ -279,6 +268,65 @@ def before_run
     COMMAND
 
     vm.sshable.cmd(command)
+    hop_configure_pgbouncer
+  end
+
+  label def configure_pgbouncer
+    pgbouncer_password = SecureRandom.urlsafe_base64(15)
+
+    pgbouncer_setup_query = <<~PGBOUNCER_SETUP
+      /**
+       * Create pgbouncer role if it does not exist.
+       */
+      DO $$
+      BEGIN
+          IF NOT EXISTS (SELECT 1 FROM pg_roles WHERE rolname = 'pgbouncer') THEN
+              CREATE ROLE pgbouncer LOGIN ENCRYPTED PASSWORD '#{pgbouncer_password}';
+          END IF;
+      END
+      $$;
+
+      /**
+       * Lock down the privileges of the pgbouncer role.
+       */
+      REVOKE ALL PRIVILEGES ON SCHEMA public FROM pgbouncer;
+
+      /**
+       * Create the pgbouncer schema if it does not exist. All of the
+       * administrative functions for pgbouncer will live in its own schema.
+       */
+      CREATE SCHEMA IF NOT EXISTS pgbouncer;
+
+      /**
+       * Lock down the privileges of the pgbouncer schema.
+       */
+      REVOKE ALL PRIVILEGES ON SCHEMA pgbouncer FROM pgbouncer;
+      GRANT USAGE ON SCHEMA pgbouncer TO pgbouncer;
+
+      /**
+       * The "get_auth" function is used by pgbouncer to authenticate users.
+       * See: http://www.pgbouncer.org/config.html#auth_query
+       */
+      CREATE OR REPLACE FUNCTION pgbouncer.get_auth (
+        INOUT p_user     name,
+        OUT   p_password text
+      ) RETURNS record
+        LANGUAGE sql SECURITY DEFINER SET search_path = pg_catalog AS
+      $$SELECT usename, passwd FROM pg_shadow WHERE usename = p_user$$;
+
+      REVOKE ALL ON FUNCTION pgbouncer.get_auth(name) FROM PUBLIC, pgbouncer;
+      GRANT EXECUTE ON FUNCTION pgbouncer.get_auth(name) TO pgbouncer;
+    PGBOUNCER_SETUP
+
+    pgbouncer_setup_result = postgres_server.run_query(pgbouncer_setup_query).chomp
+
+    raise "Failed to set up pgbouncer" unless pgbouncer_setup_result.include?("CREATE FUNCTION")
+
+    userlist_entry = "\"pgbouncer\" \"#{pgbouncer_password}\"\n"
+    vm.sshable.cmd("sudo tee /etc/postgresql/#{postgres_server.resource.version}/main/userlist.txt > /dev/null", stdin: userlist_entry)
+    vm.sshable.cmd("sudo cp /etc/postgresql/#{postgres_server.resource.version}/pgbouncer.ini /etc/pgbouncer/pgbouncer.ini")
```

Comment:
Is this a default ini file? normally I'd write the content wholesale for an application like this, but maybe my memory of the bulk in the pgbouncer.ini file is a bit hazy.

Reply:
okay I see that you are templating the whole thing downpatch, but why is it bouncing around from one location to another?

Reply:
I generally wanted to keep all generated configs for a given postgres version in `/etc/postgresql/#{postgres_server.resource.version}`,  to track pgbouncer configs for different versions. In the current state we just run a singular process using the package-provided systemd unit, so here I just copy over the config to the path it looks at for the config.

Reply:
In my experience, pgbouncer more often stands alone and independent of a postgres version. I also have implemented the multi-pgbouncer approach with systemd templates, and it worked well, other than making metrics life way more complex. It's hard for me to remember exactly, but I think I worked off this one:

https://www.enterprisedb.com/blog/running-multiple-pgbouncer-instances-systemd

Reply:
Probably worth more consideration: atomicity and syncing files, I don't know if `cp` does any of that, but `mv` is atomic and does. Anyway, I might in detail rework the config file handling, I think the coupling to postgres version might be spurious.

Reply:
This makes sense, I switched it over to directly write the config to the final location.